web/modules/comlink/src/index.ts

```diff
@@ -17,62 +22,112 @@ export type {
+
+  /**
+   * If a worker was provided, its instance will be kept here so that it
+   * can be forcefully terminated.
+   */
+  private _workerInstance?: Worker
+
   /**
    * Function intended to be used within an `extends` expression.
-   * Constructs a class that wraps around a worker factory function and
-   * automatically handles binding methods and worker creation.
+   * Constructs a class with a prototype that binds the given methods of
+   * the worker proxy. Constructors (classes) count as "methods" too.
    *
-   * @param methods - The methods to bind to the class instance, which when
-   *   called will be passed the worker instance.
+   * This function doesn't have to be used - it's just a convenience.
+   * However, you'll have to create your own methods that call out to the
+   * worker. Note that calling worker methods directly will not have any
+   * protections.
+   *
+   * @param methods - The methods to bind. These can't be figured out automatically.
    */
-  static of<T>(
-    methods: (keyof T)[]
-  ): abstract new () => AbstractWorkerBase<T> & Comlink.RemoteObject<T> {
+  static of<T>(methods: (keyof Methods<T>)[]): AbstractClass<BoundWorker<T>> {
     // @ts-ignore
-    return class extends AbstractWorkerBase<T> {
-      constructor() {
-        super()
-        for (const method of methods) {
-          // @ts-ignore
-          this[method] = async (...args: any[]) => {
-            if (this.terminated) throw new Error("Worker was already terminated!")
-            if (this.starting) await this.starting
-            if (!this.worker) await this.start()
-            if (this.methodCondition) {
-              const value = await this.methodCondition()
-              if (typeof value === "boolean" && !value) return
-            }
+    const Derived: AbstractClass<BoundWorker<T>> = class extends AbstractWorkerBase<T> {}
+
+    for (const method of methods) {
+      Derived.prototype[method] = async function (this: BoundWorker<T>, ...args: any[]) {
+        if (this.starting) await this.starting
+        if (!this.worker) await this.start()
+
+        // check one more time - maybe worker couldn't start
+        if (!this.worker) return await this._baseTryToGetDefault(method, args)
+
+        if (this._baseMethodTimeout !== 0) {
+          const result = await timedout(
             // @ts-ignore
-            const result = await this.worker[method](...args)
-            return result
-          }
+            this.worker![method](...args),
+            this._baseMethodTimeout ?? DEFAULT_TIMEOUT
+          )
+
+          if (result !== TIMED_OUT_SYMBOL) return result
+
+          // worker is timing out, have to stop it
+          this.stop()
+          throw new Error(`Method "${method}" timed out!`)
         }
       }
     }
+
+    return Derived
+  }
+
+  /** Tries to run a default method if the worker couldn't be started. */
+  private async _baseTryToGetDefault(method: keyof Methods<T>, args: any[]) {
```

Comment:
Why is this needed?

Reply:
I wanted an easy way to provide fallback values if the worker can't start. It's a bit messy to do a override function and check for this yourself, so instead I provided a specific object that you can use for providing fallbacks. This function tries to run/retrieve a fallback value for a worker method, if it exists.